src/helics/core/CoreBroker.cpp

```diff
@@ -517,18 +507,25 @@ void CoreBroker::fedRegistration(ActionMessage&& command)
     if (checkActionFlag(command, non_counting_flag)) {
         mFederates.back().nonCounting = true;
     }
+    auto lookupIndex = mFederates.size() - 1;
     if (checkActionFlag(command, child_flag)) {
         mFederates.back().global_id = GlobalFederateId(command.getExtraData());
-        mFederates.addSearchTermForIndex(mFederates.back().global_id, mFederates.size() - 1);
+        if (!mFederates.addSearchTermForIndex(mFederates.back().global_id, lookupIndex)) {
+            sendFedErrorAck(command, duplicate_federate_id);
+            return;
+        }
     } else if (isRootc) {
-        mFederates.back().global_id =
-            GlobalFederateId(static_cast<GlobalFederateId::BaseType>(mFederates.size()) - 1 +
-                             gGlobalFederateIdShift);
-        mFederates.addSearchTermForIndex(mFederates.back().global_id,
-                                         static_cast<size_t>(
-                                             mFederates.back().global_id.baseValue()) -
-                                             gGlobalFederateIdShift);
+        if (command.counter > 0 && command.counter <= 16) {
+            mFederates.back().global_id = GlobalFederateId(
+                static_cast<GlobalFederateId::BaseType>(lookupIndex) + gGlobalFederateIdShift +
+                command.counter * gGlobalPriorityBlockSize);
+        } else {
+            mFederates.back().global_id = GlobalFederateId(
+                static_cast<GlobalFederateId::BaseType>(lookupIndex) + gGlobalFederateIdShift);
+        }
+        mFederates.addSearchTermForIndex(mFederates.back().global_id, lookupIndex);
```

Comment:
Is the ID assigned to each new federate that connects just based on the current overall size of mFederates (excluding offsets due to shifts/index groups)? e.g. first federate gets ID 1, second one gets ID 2, etc?